BootloaderCommonPkg/Library/Ext23Lib/Ext2FsLs.c

```diff
@@ -0,0 +1,239 @@
+
+  Status = EFI_SUCCESS;
+  Fp = (FILE *)File->FileSystemSpecificData;
+
+  if ((Fp->DiskInode.Ext2DInodeMode & EXT2_IFMT) == EXT2_IFREG) {
+    DEBUG ((DEBUG_INFO, "  %a\t%d\n", File->FileNamePtr, Fp->DiskInode.Ext2DInodeSize));
+    return EFI_SUCCESS;
+  } else if ((Fp->DiskInode.Ext2DInodeMode & EXT2_IFMT) != EXT2_IFDIR) {
+    return EFI_NOT_FOUND;
+  }
+
+  BlockSize = Fp->SuperBlockPtr->Ext2FsBlockSize;
+  Names = NULL;
+  Fp->SeekPtr = 0;
+  while (Fp->SeekPtr < (OFFSET)Fp->DiskInode.Ext2DInodeSize) {
+    Rc = BufReadFile (File, &Buf, &BufSize);
+    if (Rc != 0) {
+      Status = EFI_DEVICE_ERROR;
+      goto out;
+    }
+    if ((BufSize != BlockSize) || (BufSize == 0)) {
+      Status = EFI_DEVICE_ERROR;
+      goto out;
+    }
+
+    Dp = (EXT2FS_DIRECT *)Buf;
+    EdPtr = (EXT2FS_DIRECT *)(Buf + BufSize);
+
+    for (; Dp < EdPtr;  Dp = (VOID *) ((CHAR8 *)Dp + Dp->Ext2DirectRecLen)) {
+      if (Dp->Ext2DirectRecLen <= 0) {
+        Status = EFI_DEVICE_ERROR;
+        goto out;
+      }
+
+      if (Dp->Ext2DirectInodeNumber == 0) {
+        continue;
+      }
+
+      if (Dp->Ext2DirectType >= NELEM (mTypeStr)) {
+        //
+        //  This does not handle "Old"
+        //  filesystems properly. On little
+        //  endian machines, we get a bogus
+        //  type Name if the NameLen matches a
+        //  valid type identifier. We could
+        //  check if we read NameLen "0" and
+        //  handle this case specially, if
+        //  there were a pressing need...
+        //
+        DEBUG ((DEBUG_INFO, "bad dir entry - %d\n", Dp->Ext2DirectType));
+        Status = EFI_DEVICE_ERROR;
+        goto out;
+      }
+      Type = mTypeStr[Dp->Ext2DirectType];
+
+      New = AllocateZeroPool (sizeof * New + AsciiStrLen (Dp->Ext2DirectName));
+      if (New == NULL) {
+        DEBUG ((DEBUG_INFO, "%d: %s (%s)\n",
+                Dp->Ext2DirectInodeNumber, Dp->Ext2DirectName, Type));
+        continue;
+      }
+
+      New->EntryInode = Dp->Ext2DirectInodeNumber;
+      New->EntryType  = Dp->Ext2DirectType;
+      AsciiStrCpyS (New->EntryName, EXT2FS_MAXNAMLEN, Dp->Ext2DirectName);
+      New->EntryName[Dp->Ext2DirectNameLen] = '\0';
+      for (NextPtr = &Names; *NextPtr != NULL; NextPtr = & (*NextPtr)->EntryNext) {
+          if (AsciiStrCmp (New->EntryName, (*NextPtr)->EntryName) < 0) {
+          break;
+        }
+      }
+      New->EntryNext = *NextPtr;
+      *NextPtr = New;
+    }
+    Fp->SeekPtr += BufSize;
+  }
+
+  if (Names != NULL) {
+    PNames = Names;
+
+    do {
+      New = PNames;
+      if (New->EntryType == 2) {
+        DEBUG ((DEBUG_INFO, "  %a/\n", New->EntryName));
+      } else if (New->EntryType == 1) {
+        FileSize = 0;
+        Rc = ReadInode (New->EntryInode, File);
+        if (!Rc) {
+          Fp = (FILE *)File->FileSystemSpecificData;
+          FileSize = Fp->DiskInode.Ext2DInodeSize;
+        }
+        DEBUG ((DEBUG_INFO, "  %a\t%d\n", New->EntryName, FileSize));
```

Comment:
For release build,  what is the behavior ?
Since the display console and serial console will see different result,  will it cause confusion ?
Can we consider returning a text buffer, and later on we use ShellPrint to output ?

Reply:
Right, I missed the display console support. To support it, dir/file entries needs to be passed to shell. A buffer or linked list needs to be allocated in each FS library, and caller(shell) may need to de-allocate the memory unless another cleanup api is provided. Is it okay to support display console in a separate PR?